overlay/usr/local/etc/nginx/conf.d/nextcloud.inc

```diff
@@ -0,0 +1,125 @@
+
+# Pagespeed is not supported by Nextcloud, so if your server is built
+# with the `ngx_pagespeed` module, uncomment this line to disable it.
+#pagespeed off;
+
+# HTTP response headers borrowed from Nextcloud `.htaccess`
+add_header Referrer-Policy                      "no-referrer"   always;
+add_header X-Content-Type-Options               "nosniff"       always;
+add_header X-Download-Options                   "noopen"        always;
+add_header X-Frame-Options                      "SAMEORIGIN"    always;
+add_header X-Permitted-Cross-Domain-Policies    "none"          always;
+add_header X-Robots-Tag                         "none"          always;
+add_header X-XSS-Protection                     "1; mode=block" always;
+
+# Remove X-Powered-By, which is an information leak
+fastcgi_hide_header X-Powered-By;
+
+# Path to the root of your installation
+root /usr/local/www/nextcloud/;
+
+# Specify how to handle directories -- specifying `/index.php$request_uri`
+# here as the fallback means that Nginx always exhibits the desired behaviour
+# when a client requests a path that corresponds to a directory that exists
+# on the server. In particular, if that directory contains an index.php file,
+# that file is correctly served; if it doesn't, then the request is passed to
+# the front-end controller. This consistent behaviour means that we don't need
+# to specify custom rules for certain paths (e.g. images and other assets,
+# `/updater`, `/ocm-provider`, `/ocs-provider`), and thus
+# `try_files $uri $uri/ /index.php$request_uri`
+# always provides the desired behaviour.
+index index.php index.html /index.php$request_uri;
+
+# Rule borrowed from `.htaccess` to handle Microsoft DAV clients
+location = / {
+    if ( $http_user_agent ~ ^DavClnt ) {
+        return 302 /remote.php/webdav/$is_args$args;
+    }
+}
+
+location = /robots.txt {
+    allow all;
+    log_not_found off;
+    access_log off;
+}
+
+# Make a regex exception for `/.well-known` so that clients can still
+# access it despite the existence of the regex rule
+# `location ~ /(\.|autotest|...)` which would otherwise handle requests
+# for `/.well-known`.
+location ^~ /.well-known {
+    # The rules in this block are an adaptation of the rules
+    # in `.htaccess` that concern `/.well-known`.
+
+    location = /.well-known/carddav { return 301 $scheme://$host:$NEXTCLOUD_HTTPS_PORT/remote.php/dav/; }
+    location = /.well-known/caldav  { return 301 $scheme://$host:$NEXTCLOUD_HTTPS_PORT/remote.php/dav/; }
+
+    location /.well-known/acme-challenge    { try_files $uri $uri/ =404; }
+    location /.well-known/pki-validation    { try_files $uri $uri/ =404; }
+
+    # Let Nextcloud's API for `/.well-known` URIs handle all other
+    # requests by passing them to the front-end controller.
+    return 301 $scheme://$host:$NEXTCLOUD_HTTPS_PORT/index.php$request_uri;
+}
```

Comment:
I think this block may have to have some tweaks to support both HTTPS and non-HTTPS (it references the HTTPS port env variable).

Reply:
This is a redirect, so it has to be HTTPS as Nextcloud is, in the end, accessed through HTTPS.

However, this line expects the external proxy to use the same port as the internal one, which might not be the case.

Reply:
Relative redirects are probably the solution here:

https://stackoverflow.com/questions/45396651/nginx-redirect-without-scheme-and-hostname
https://stackoverflow.com/questions/33523821/how-to-issue-a-relative-url-redirect-from-nginx

This would essentially result in adding this somewhere in the nextcloud server block:

```nginx
    absolute_redirect off;
```

I verified that this works, and relative URIs are valid for the `Location:` header in redirects in [RFC7231](https://www.rfc-editor.org/rfc/rfc7231#section-7.1.2). 

This might help the other redirects for nextcloud as well (like Microsoft DAV client redirect above).

Reply:
Was seen :).

Weirdly this worked on the first attempt whereas I couldn't make the `absolute_redirect` directive work last time that I tried. Hopefully, I am not missing another constraint.

Reply:
That's odd, it's working great for me on two different nextcloud instances behind a reverse proxy.

What wasn't working about it the last time you tried it?

Reply:
Don't remember exactly, but I think `absolute_redirect off` was simply not making the redirect relative.

Reply:
You also changed the redirects themselves to not have the `scheme://host:port` part? If you were testing in a browser, make sure to disable the cache.

For me, internally:
```sh
$ wget "http://10.1.0.24/.well-known/carddav" 2>&1 | grep -E '(Location:|http)'
--2021-11-17 11:00:35--  http://10.1.0.24/.well-known/carddav
Location: /remote.php/dav/ [following]
--2021-11-17 11:00:35--  http://10.1.0.24/remote.php/dav/
```

and externally:

```sh
$ wget "https://<host.fqdn>/.well-known/carddav" 2>&1 | grep -E '(Location:|http)'
--2021-11-17 11:00:35--  https://<host.fqdn>/.well-known/carddav
Location: /remote.php/dav/ [following]
--2021-11-17 11:00:35--  https://<host.fqdn>/remote.php/dav/
```

Reply:
Didn't mean to resolve this one, yet, oops.